Better interface for custom node storage (Draft 2)

bigerl opened the pull request:
this supersedes #21

bigerl committed:
use std::string_view instead of std::string &const

bigerl committed:
(possible) fix for bad comparison performance

bigerl committed:
add option to manually register and unregister backends

bigerl committed:
fixed typo

bigerl committed:
[WIP] messenger objects to abstract away Backend storage implementations

bigerl committed:
[WIP] extracted BackendNodeHandles.hpp

bigerl committed:
[WIP] use std::string_view as constructor arguments in classes derived from Node

bigerl committed:
stabilizing portable node storage backends

bigerl committed:
NodeID, NodeBackendHandle and Node noexcept for improved performance in collections

bigerl committed:
fixed imports

bigerl committed:
avoid NodeID copy

bigerl committed:
node id is fully no-throw

bigerl committed:
draft 2, first compilable version

bigerl committed:
documentation and renaming.

bigerl committed:
fixed typo

bigerl commented on src/rdf4cpp/rdf/storage/node/NodeStorage.hpp:
```diff
@@ -17,88 +17,332 @@ namespace rdf4cpp::rdf::storage::node {
+ * This does not apply to the default_instance.
+ * The live cycle of NodeStorage's and their backends is managed by the static methods within this class.
+ */
 class NodeStorage {
```
@nkaralis I tried to address the points from your review https://github.com/rdf4cpp/rdf4cpp/pull/21#issuecomment-1025506011 . Could you check if you think that this reworked interface is an improvement and reevaluate?

nkaralis replied:
The split of the functions looks good to me.

bigerl commented:
Things to address before merge:

- [ ] make sure null() NodeID's are nowhere used (could be that they are currently used for default graph IRIs).
- [ ] fix race condition in INodeStorageBackend
- [ ] document INodeStorageBackend
- [ ] document ReferenceNodeStorageBackend
- [ ] document Views
- [ ] document identifier::*

bigerl committed:
documentation for identifiers

bigerl committed:
rename NodeBackendHandle empty to null

bigerl committed:
INodeStorageBackend synchronization and documentaiton

bigerl committed:
improved LiteralBackend.cpp quote_lexical

bigerl committed:
more documentation for Identifiers

bigerl committed:
more documentation

bigerl committed:
added documentation

bigerl committed:
documentation and final rename namespace handle to view

bigerl committed:
fixing NodeStorage constructors

nkaralis commented:
I have a few comments on the new additions. 

- First of all, I found the documentation very helpful. 
- Regarding NodeStorageID, I think its documentation has conflicts. The `identifier/README.md` states that NodeStorageID is used to identify the NodeStorage of a Node. However, the file `NodeStorageID.hpp` states that a NodeStorageID identifies a Literal of a given type within a given NodeStorage. Is the usage of NodeStorageID limited only to Literals?
- In `ReferenceBackendImplementation.hpp/.cpp` the functions `find_*_backend_view()` are marked `noexcept`, however they use `std::map::at method`, which throws an exception if the provided key is missing. Would it be better to first check with `std::map::contains` if the provided key exists? In case the provided key is missing, we could return a `nullptr`

bigerl committed:
fixed NodeStorageID.hpp docstring

bigerl commented:
> I have a few comments on the new additions.
> 
>     * First of all, I found the documentation very helpful.
> 
>     * Regarding NodeStorageID, I think its documentation has conflicts. The `identifier/README.md` states that NodeStorageID is used to identify the NodeStorage of a Node. However, the file `NodeStorageID.hpp` states that a NodeStorageID identifies a Literal of a given type within a given NodeStorage. Is the usage of NodeStorageID limited only to Literals?
> 
>     * In `ReferenceBackendImplementation.hpp/.cpp` the functions `find_*_backend_view()` are marked `noexcept`, however they use `std::map::at method`, which throws an exception if the provided key is missing. Would it be better to first check with `std::map::contains` if the provided key exists? In case the provided key is missing, we could return a `nullptr`

I've addressed all points from this review. `find_*_backend_view` are in the current code not noexcept but are expected to throw an exception if the id does not exist. This is documented in the interface definition at INodeStorageBackend.hpp.
